fdk.go

```diff
@@ -3,94 +3,142 @@ package fdk
 import (
 	"context"
 	"io"
+	"log"
 	"net/http"
 	"os"
-
-	"github.com/fnproject/fdk-go/utils"
+	"time"
 )
 
+// Handler is a function handler, representing 1 invocation of a function
 type Handler interface {
-	Serve(ctx context.Context, in io.Reader, out io.Writer)
+	// Serve contains a context with request configuration, the body of the
+	// request as a stream of bytes, and a writer to output to; user's may set
+	// headers via the resp writer using the fdk's SetHeader/AddHeader methods -
+	// if you've a better idea, pipe up.
+	Serve(ctx context.Context, body io.Reader, resp io.Writer)
 }
 
+// HandlerFunc makes a Handler so that you don't have to!
 type HandlerFunc func(ctx context.Context, in io.Reader, out io.Writer)
 
+// Serve implements Handler
 func (f HandlerFunc) Serve(ctx context.Context, in io.Reader, out io.Writer) {
 	f(ctx, in, out)
 }
 
-// Context will return an *fn.Ctx that can be used to read configuration and
+// GetContext will return an fdk Context that can be used to read configuration and
 // request information from an incoming request.
-func Context(ctx context.Context) *Ctx {
-	utilsCtx := utils.Context(ctx)
-	return &Ctx{
-		HTTPHeader: utilsCtx.HTTPHeader,
-		Header:     utilsCtx.Header,
-		Config:     utilsCtx.Config,
-		RequestURL: utilsCtx.RequestURL,
-		Method:     utilsCtx.Method,
-	}
+func GetContext(ctx context.Context) Context {
+	return ctx.Value(ctxKey).(Context)
 }
 
-func WithContext(ctx context.Context, fnctx *Ctx) context.Context {
-	utilsCtx := &utils.Ctx{
-		HTTPHeader: fnctx.HTTPHeader,
-		Header:     fnctx.Header,
-		Config:     fnctx.Config,
-		RequestURL: fnctx.RequestURL,
-		Method:     fnctx.Method,
-	}
-	return utils.WithContext(ctx, utilsCtx)
+// WithContext adds an fn context to a context context. It is unclear why this is
+// an exported method but hey here ya go don't hurt yourself.
+func WithContext(ctx context.Context, fnctx Context) context.Context {
+	return context.WithValue(ctx, ctxKey, fnctx)
+}
+
+type key struct{}
+
+var ctxKey = new(key)
+
+// Context contains all configuration for a function invocation
+type Context interface {
+	// Config is a map of all env vars set on a function, the base set of fn
+	// headers in addition to any app and function configuration.
+	Config() map[string]string
+
+	// Header are the headers sent to this function invocation
+	Header() http.Header
+
+	// ContentType is Header().Get("Content-Type") but with 15 less chars, you are welcome
+	// XXX(discuss): is this necessary? Header.Get("Content-Type") is easy enough for a caveman...
+	ContentType() string
+
+	// CallID is the call id for this function invocation
+	CallID() string
+
+	// TODO should we unwind all known env var fields as methods here, as well?
```

Comment:
I don't really think it matters either way - in other FDK's i've added these just for convenience but it's a choice I think